src/Orleans.Core.Abstractions/IDs/GrainIdKeyExtensions.cs

```diff
@@ -12,37 +13,84 @@ public static class GrainIdKeyExtensions
         /// <summary>
         /// Creates an <see cref="IdSpan"/> representing a <see cref="long"/> key.
         /// </summary>
-        public static IdSpan CreateIntegerKey(long key) => IdSpan.Create(key.ToString("X"));
+        public static IdSpan CreateIntegerKey(long key)
+        {
+            Span<byte> buf = stackalloc byte[sizeof(long) * 2];
+            Utf8Formatter.TryFormat(key, buf, out var len, 'X');
+            return new IdSpan(buf.Slice(0, len).ToArray());
+        }
 
         /// <summary>
         /// Creates an <see cref="IdSpan"/> representing a <see cref="long"/> key.
         /// </summary>
-        public static IdSpan CreateIntegerKey(long key, string keyExtension) => string.IsNullOrWhiteSpace(keyExtension) ? CreateIntegerKey(key) : IdSpan.Create($"{key:X}+{keyExtension}");
+        public static IdSpan CreateIntegerKey(long key, string keyExtension)
+        {
+            if (string.IsNullOrWhiteSpace(keyExtension)) return CreateIntegerKey(key);
+
+            Span<byte> tmp = stackalloc byte[sizeof(long) * 2];
+            Utf8Formatter.TryFormat(key, tmp, out var len, 'X');
+
+            var extLen = Encoding.UTF8.GetByteCount(keyExtension);
+            var buf = new byte[len + 1 + extLen];
+            tmp.Slice(0, len).CopyTo(buf);
+            buf[len] = (byte)'+';
+            Encoding.UTF8.GetBytes(keyExtension, 0, keyExtension.Length, buf, len + 1);
+
+            return new IdSpan(buf);
+        }
 
         /// <summary>
         /// Creates an <see cref="IdSpan"/> representing a <see cref="Guid"/> key.
         /// </summary>
-        public static IdSpan CreateGuidKey(Guid key) => IdSpan.Create(key.ToString("N"));
+        public static IdSpan CreateGuidKey(Guid key)
+        {
+            var buf = new byte[32];
+            Utf8Formatter.TryFormat(key, buf, out _, 'N');
+            return new IdSpan(buf);
+        }
 
         /// <summary>
         /// Creates an <see cref="IdSpan"/> representing a <see cref="Guid"/> key.
         /// </summary>
-        public static IdSpan CreateGuidKey(Guid key, string keyExtension) => string.IsNullOrWhiteSpace(keyExtension) ? CreateGuidKey(key) : IdSpan.Create($"{key:N}+{keyExtension}");
+        public static IdSpan CreateGuidKey(Guid key, string keyExtension)
+        {
+            if (string.IsNullOrWhiteSpace(keyExtension)) return CreateGuidKey(key);
+
+            var extLen = Encoding.UTF8.GetByteCount(keyExtension);
+            var buf = new byte[32 + 1 + extLen];
+            Utf8Formatter.TryFormat(key, buf, out _, 'N');
```

Comment:
Perhaps there should at least be a Debug.Assert on the result of these operations, if not throwing an exception if they fail - despite us being sure that it will succeed given the buffer size is pre-determined.